std/net/udp.zig

```diff
@@ -0,0 +1,121 @@
+const std = @import("../index.zig");
+const event = std.event;
+const assert = std.debug.assert;
+const os = std.os;
+const mem = std.mem;
+const net = std.net;
+
+pub fn UDP() type {
+    return struct {
+        channel: *event.Channel(Event),
+        putter: promise,
+        table_lock: event.Lock,
+        fd: i32,
+
+        const Self = this;
+
+        pub const Event = union(enum) {
+            Packet: []u8,
+            Err: Error,
+
+            pub const Error = error{
+                UserResourceLimitReached,
+                SystemResources,
+            };
+        };
+
+        pub fn listen(address: net.Address, loop: *event.Loop, event_buf_count: usize) !*Self {
+            //socket
+            const fd = try os.posixSocket(address.family(), os.posix.SOCK_DGRAM, os.posix.IPPROTO_UDP);
+            errdefer os.close(fd);
+
+            //set non blocking?
+
+            //bind
+            //@breakpoint();
+            try os.posixBind(fd, address);
+
+            const channel = try event.Channel(Self.Event).create(loop, event_buf_count);
+            errdefer channel.destroy();
+
+            var result: *Self = undefined;
+            _ = try async<loop.allocator> eventPutter(fd, channel, &result);
+            return result;
+        }
+
+        pub fn destroy(self: *Self) void {
+            cancel self.putter;
+        }
+
+        pub async fn write(self: *Self, file_path: []const u8, value: V) !?V {
+        }
+
+        async fn eventPutter(fd: i32, channel: *event.Channel(Event), out: **Self) void {
+            // TODO https://github.com/ziglang/zig/issues/1194
+            suspend {
+                resume @handle();
+            }
+
+            const loop = channel.loop;
+
+            var udp = Self{
+                .putter = @handle(),
+                .channel = channel,
+                .table_lock = event.Lock.init(loop),
+                .fd = fd,
+            };
+            out.* = &udp;
+
+            loop.beginOneEvent();
+
+            defer {
+                loop.finishOneEvent();
+                os.close(fd);
+                channel.destroy();
+            }
+
+            var event_buf: [1500]u8 = undefined;
+
+            while (true) {
+                const rc = os.darwin.read(fd, &event_buf, event_buf.len);
```

Comment:
You need to use `recvmsg()` here, and the buf above should support jumbo frames (9000 mtu). On Linux you should use `recvmmsg()`. I don't think a buf of exactly 9000 is correct with `recvmsg()` because the ip header stuff is not included.

Reply:
You are correct. This is what I currently have locally:

```
+            const rc = os.posix.recvfrom( ev_handle
                                        , &event_buf
                                        , event_buf.len
                                        , 0
                                        , &recv_addr.sa.u.sa
                                        , &recv_addr.sa.len
                                        );

            const errno = os.posix.getErrno(rc);
            switch (errno) {
                0 => {
                    std.debug.warn("PACKET: {}\n", &event_buf);
                    await (async channel.put(Self.Event{
                        .id = UDPEventId.Packet,
                        .address = recv_addr,
                        //.packet = &event_buf,
                    }) catch unreachable);
                },
                os.posix.EINTR => continue,
                os.posix.EINVAL => unreachable,
                os.posix.EFAULT => unreachable,
                // TODO this needs to try to re-up the Handler for the socket
                // iPhone is known for doing throwing this when a handoff occurs
                os.posix.ENOTCONN => unreachable,
                os.posix.EAGAIN => {
                    _ = (await (async loop.waitEvHandle(
                        ev_handle,
                        // https://github.com/ziglang/zig/issues/485
                        event.Loop.EventFlags { .READ = true
                                              , .WRITE = false
                                              , .EXCEPT = false
                                              , .reserved = undefined },
                    ) catch unreachable)) catch |err| {
                        const transformed_err = switch (err) {
                            error.AccessDenied,
                            error.EventNotFound,
                            error.ProcessNotFound,
                            error.OutOfMemory => unreachable,
                            error.SystemResources => UDPEventError.SystemResources,
                        };
                        await (async channel.put(transformed_err) catch unreachable);
                        return;
                    };
                },
                else => unreachable,
```